### ISSUE TITLE
[Feature Request]: Add clipboard Clearing and Switch User

### ISSUE THREAD

**Issue by Enigma1309** [prasanthrangan/hyprdots#709]
# 1st
- I want the clipboard to be cleared, either with a keybind or at startup and also the ability to delete certain things from clipboard,
 Say i have 3 items 
-- 1 [python NaSa HacKinG script]
-- 2 [Mail Password]
-- 3  [Image]
i want to delete the second item of the clipboard list, so add a button to delete it.
after delete it should look like this
-- 1 [python NaSa HacKinG script]
-- 2 [Image]

# 2nd
- Add ability to switch user, I have two users in my pc i usually hop between these two so add an ability to switch between users rather than logging out.

Some thing like this (yeah i don't have good photo, editing experience)

![image](https://github.com/prasanthrangan/hyprdots/assets/94917588/a2e4e3db-3695-4e54-a531-4dde72fcdc81)

**Comment by Enigma1309** [comment]
Also if it's possible to add a Notification center , like gnome has.. it would be and absolute win-win

**Comment by kRHYME7** [comment]
1.  You can add this to your keybind
```
❯ ~/.config/hypr/scripts/cliphist.sh
cliphist.sh [action]
c :  cliphist list and copy selected
d :  cliphist list and delete selected
w :  cliphist wipe database
```

2. Where can we find that feature do you have links for that dots ?

**Comment by Enigma1309** [comment]
Idk any dots that has been having effective clipboard support but i think windows have this natively
![image](https://github.com/prasanthrangan/hyprdots/assets/94917588/74797672-5e63-46ea-a5c1-47a04f8d9990)

yeah ik we don't wanna copy windows here, but it would give a better functionality.

For notification you can see the dots files from follwing repo:
- [this uses waybar](https://github.com/momcilovicluka/Hyprland-dots)
- [this one is very beautiful, but uses eww](https://github.com/Aylur/dotfiles)
- Just search for more on github you'll find plenty of configs

---
> You can add this to your keybind

Secondly i already tried it but it isn't working (shows no popup)
![image](https://github.com/prasanthrangan/hyprdots/assets/94917588/f752ba8c-92ff-43d2-89b4-f45061058bd2)

**Comment by Enigma1309** [comment]
I have two DE setup in the machine, Gnome (for family) and Hyprland.
i can easily switch user from gnome to hyprland without actually logging out but i can't do the other way round, i've to logout in order to switch to other user.
(also i'm using gdm instead of sddm, if that counts..)

**Comment by kRHYME7** [comment]
> Secondly i already tried it but it isn't working (shows no popup)

About the clipboard, won't this show up?  How about on the terminal? (I guess we can add a mark feature on the clipboard to easily select and delete)
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/df0be879-ebbb-42c4-adce-136ade29cc66)


>yeah ik we don't wanna copy windows here, but it would give a better functionality.

This would be very nice but as of rofi's capability, we can't go further (I guess). but maybe we can utilized eww or ags for making a clipboard widget, or some other toolkit (I hope someone has already made it work)  that can be implemented for these dots 



>[this uses waybar](https://github.com/momcilovicluka/Hyprland-dots)
[this one is very beautiful, but uses eww](https://github.com/Aylur/dotfiles)
Just search for more on github you'll find plenty of configs

The first link was swaync and I think this is easy to implement soon. (Wll try to do this)
the last link was ags,  I guess we can use that, as the language it uses is common. But it exhibited bugs a few months ago. Will try this out soon, I hope someone can initiate this. (Needs extra effort )



>I have two DE setup in the machine, Gnome (for family) and Hyprland.
i can easily switch user from gnome to hyprland without actually logging out but i can't do the other way round, i've to logout in order to switch to other user.
(also i'm using gdm instead of sddm, if that counts..)

Will see some configs that may have these features and make it work on hyprdots.

Those features are for sure handy. Hoping for this to be implemented soon. I'll try to draft something too regarding this Features.

**Comment by Enigma1309** [comment]
> About the clipboard, won't this show up? How about on the terminal? (I guess we can add a mark feature on the clipboard to easily select and delete)

can you send the code it's not working in my case 

> Those features are for sure handy. Hoping for this to be implemented soon.

thank you so much for the help i hope it comes to action soon <3

**Comment by kRHYME7** [comment]
https://github.com/prasanthrangan/hyprdots/blob/main/Configs/.config/hypr/scripts/cliphist.sh

Here's the link it should be the same on your system. 

Also what is the output when you run it on the terminal?

**Comment by Enigma1309** [comment]
> https://github.com/prasanthrangan/hyprdots/blob/main/Configs/.config/hypr/scripts/cliphist.sh
> 
> Here's the link it should be the same on your system.

Yes it's same on my system, but doesn't work through keybind tho.

> Also what is the output when you run it on the terminal?

It's working fine with terminal, jut like you're, but not working with keybind


### ohh wait it worked, i wrote my keybinds into keybindings.conf but it wasn't working, i wrote it in userprefs.conf, now it works 
probably somewhere it was overwritten.

thanks for the help, this problem is solved partially, i hope for the better implementation in future

**Comment by kRHYME7** [comment]
I think you should write it in ``` ~/.config/hypr/keybindings.conf ``` to take effect. 

you can run ``` Hyprdots show binds ``` to see what binds are active

Also you can  add  ```#Description  ```
To have a beautiful name for your exec keybinds

Without comments: 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/6bc8ae9e-4b4e-45fc-8efe-984f23b2037d)


# Added some Description 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/9ae73734-48d4-43f1-aa54-e3ad255d66c1)

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/9b596e62-61b6-4c93-870b-3cdeb676f7db)

**Comment by Enigma1309** [comment]
> ❯ Hyprdots show binds
> jq: error: syntax error, unexpected IDENT, expecting '}' (Unix shell quoting issues?) at /tmp/hyprdots-keybinds.jq, line 45:
> "kitty --title "kitty_fly" btop" : "kitty --title "kitty_fly" btop",                
> jq: error: May need parentheses around object key expression at /tmp/hyprdots-keybinds.jq, line 4:
> "~/.config/hypr/scripts/dontkillsteam.sh" : "Killactive",
> jq: error: syntax error, unexpected IDENT, expecting '}' (Unix shell quoting issues?) at /tmp/hyprdots-keybinds.jq, line 45:
> "kitty --title "kitty_fly" btop" : "kitty --title "kitty_fly" btop",                                                   
> jq: 3 compile errors

Probably this line in userprefs.conf is creating issue.
![image](https://github.com/prasanthrangan/hyprdots/assets/94917588/1e0b04e6-dde5-4a8c-9bb0-6c4b0fba91e6)

I made this to make kitty windows un-tiled when launching

![image](https://github.com/prasanthrangan/hyprdots/assets/94917588/5dd5b6df-507b-457b-be52-08e75b79477c)

This is my window rule


IDK why it's not working in keybinds.conf but only in userprefs.conf

**Comment by Enigma1309** [comment]
Okay that issue is sorted...now it's working in keybinds.conf file... IDK how it fixed but it worked automatically... Thanks

**Comment by kRHYME7** [comment]
I'm curious about which dotfiles have this configuration. 

![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/1f5989b3-76eb-40b5-8436-9adf10b1c5b3)

**Comment by Enigma1309** [comment]
I haven't seen any dotfiles which has switch user feature (as of now) , idk how it works, this was just my attempt of how it might look if we add something like this in our configuration 🙃

---

BTW what about Notification Center isn't anyone interested in making it work here? 🙄

**Comment by kRHYME7** [comment]
> BTW what about Notification Center isn't anyone interested in making it work here? 🙄

# Pointers

Updates: 
Trying to implement the colorchanging based on wallpaper implementations. (Super not yet)
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/a62bef38-06d7-43e2-8796-c9b9bc7610f8)

If we only want the control center of swaync we should just disable the notification part and add widgets capabilities for swaync's notification panel.  In short swaync won't handle any notifications

Why???
Problems and Conflicts: 

If we want a full functionality of swaync  we need some editing of the scripts which uses dunst (the notification daemon used in Hyprdots). 
(Which I guess this should be in supervision with @prasanthrangan as I am afraid that this will have a bunch of file changes), this process will just be changing the command ``` dunstify ``` to more generic command ``` notify-send ```  reference at the bottom, dunstify and notify-send commands are pretty similar though. 
However, issues like this occurs if we use swaync as the notification daemon (Unless rewriting those to cater swaync )
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/e6b20311-9fa6-4f0a-a032-5fa72b509f9e)



Best solution is fork this repo and use swaync instead of dunst. 



















## Reference:
```
❯ notify-send --help
Usage:
  notify-send [OPTION…] <SUMMARY> [BODY] - create a notification

Help Options:
  -?, --help                        Show help options

Application Options:
  -u, --urgency=LEVEL               Specifies the urgency level (low, normal, critical).
  -t, --expire-time=TIME            Specifies the timeout in milliseconds at which to expire the notification.
  -a, --app-name=APP_NAME           Specifies the app name for the icon
  -i, --icon=ICON                   Specifies an icon filename or stock icon to display.
  -c, --category=TYPE[,TYPE...]     Specifies the notification category.
  -e, --transient                   Create a transient notification
  -h, --hint=TYPE:NAME:VALUE        Specifies basic extra data to pass. Valid types are boolean, int, double, string, byte and variant.
  -p, --print-id                    Print the notification ID.
  -r, --replace-id=REPLACE_ID       The ID of the notification to replace.
  -w, --wait                        Wait for the notification to be closed before exiting.
  -A, --action=[NAME=]Text...       Specifies the actions to display to the user. Implies --wait to wait for user input. May be set multiple times. The name of the action is output to stdout. If NAME is not specified, the numerical index of the option is used (starting with 0).
  -v, --version                     Version of the package.

❯ dunstify --help
Usage:
  dunstify [OPTION…] SUMMARY BODY

Help Options:
  -?, --help                  Show help options

Application Options:
  -a, --appname=NAME          Name of your application
  -u, --urgency=URG           The urgency of this notification
  -h, --hints=HINT            User specified hints
  -A, --action=ACTION         Actions the user can invoke
  -t, --timeout=TIMEOUT       The time in milliseconds until the notification expires
  -i, --icon=ICON             An Icon that should be displayed with the notification
  -I, --raw_icon=RAW_ICON     Path to the icon to be sent as raw image data
  -c, --capabilities          Print the server capabilities and exit
  -s, --serverinfo            Print server information and exit
  -p, --printid               Print id, which can be used to update/replace this notification
  -r, --replace=ID            Set id of this notification.
  -C, --close=ID              Close the notification with the specified ID
  -b, --block                 Block until notification is closed and print close reason

❯ swaync-client --help


Usage:
  swaync-client <OPTION>
Help:
  -h, 	--help 			Show help options
  -v, 	--version 			Prints version
Options:
  -R, 	--reload-config 		Reload the config file
  -rs, 	--reload-css 			Reload the css file. Location change requires restart
  -t, 	--toggle-panel 		Toggle the notificaion panel
  -op, 	--open-panel 			Opens the notificaion panel
  -cp, 	--close-panel 			Closes the notificaion panel
  -d, 	--toggle-dnd 			Toggle and print the current dnd state
  -D, 	--get-dnd 			Print the current dnd state
  -dn, 	--dnd-on 			Turn dnd on and print the new dnd state
  -df, 	--dnd-off 			Turn dnd off and print the new dnd state
  -I, 	--get-inhibited 		Print if currently inhibited or not
  -In, 	--get-num-inhibitors 		Print number of inhibitors
  -Ia, 	--inhibitor-add [APP_ID] 	Add an inhibitor
  -Ir, 	--inhibitor-remove [APP_ID] 	Remove an inhibitor
  -Ic, 	--inhibitors-clear 		Clears all inhibitors
  -c, 	--count 			Print the current notificaion count
      	--hide-latest 			Hides latest notification. Still shown in Control Center
      	--close-latest 		Closes latest notification
  -C, 	--close-all 			Closes all notifications
  -sw, 	--skip-wait 			Doesn't wait when swaync hasn't been started
  -s, 	--subscribe 			Subscribe to notificaion add and close events
  -swb,  --subscribe-waybar 		Subscribe to notificaion add and close events with waybar support. Read README for example

```

**Comment by kRHYME7** [comment]
> I haven't seen any dotfiles which has switch user feature (as of now) , idk how it works, this was just my attempt of how it might look if we add something like this in our configuration 🙃



If you meant how gnome/kde changes users.
If I understand file systems, Here'e the thing if it is possible. (Too complicated for user dotfiles level as this is already a $USER configuration)

For now, switching $USER1 from this dot files to other $USER2 means that after you Switch user from this the dotfiles won't be retained for $USER2 as you can see the home directory of $USER2 will have a different .config. This means you will maintain 2 .configs.  Unless this user dot files are a symbolic links from example : ```  /usr/local/share  ```, but with this $USER1 and $USER2 will always have the same configuration. And so on ....

And I guess this should be in Discussion panel though

**Comment by kRHYME7** [comment]
@Enigma1309 
This is the only I can do with minimal changes to the clipboard. Added --multi flag to have multi-selection also not showing the entry numbers  

cons:
all lines will have ```,,,```


running: ``` Hyprdots show pasteboard d --multi ```
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/e8a94d34-2afc-46e3-ae95-38f2691b1cee)

without --multi
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/f72ade47-7945-4e11-a237-b9696486de4f)


I hosted it here first to just to see if someone might want to use this feature also I'll be issuing a PR if majority wants this feature

**Comment by Enigma1309** [comment]
> Hyprdots show pasteboard d --multi

Okay this is working, also i have to press shift key to select multiple maybe that's intensional. Could we change it to CTRL key for multiple and use shift key for selecting a range??



> If you meant how gnome/kde changes users.

Hey maybe we can try to change TTY to another one? and it will launch Login manager??maybe...?

>  I hope someone can manage to create it.

if anyone manages to create it then please consider mentioning/commenting or posting a link....

**Comment by kRHYME7** [comment]
> Okay this is working, also i have to press shift key to select multiple maybe that's intensional. Could we change it to CTRL key for multiple and use shift key for selecting a range??

This should be a user-specific change and it is under the scope of rofi. let's try to find a way to change this as this is a rofi thing and this bind is just the default.

>Hey maybe we can try to change TTY to another one? and it will launch Login manager??maybe...?

Hmm Is this  just like logging out then login again to a different user? If you want I guess you can Just ``` ctrl + alt + 2 ``` then this will I guess show sddm? then  try to login. I will try to test this as I am not using sddm.

**Comment by Enigma1309** [comment]
TL;DR :- Switch to TTY 1 for technically switching user without loggin out 

i think i found the solution....whenever we start the machine the login manager sits on the TTY 1 whenever we login it send us to another free tty, so it means that the login manager is never closing and just redirects to another tty and does the login process.
So what we can do is to write a script that switches us to  TTY 1 (login manager) in the process of switching we are never being actually logged out... and the fun fact is that if we have already logged in then the login manager will redirect us to the specific tty which is already logged in (instead of creating new). I think the major problem is solved. Now we just have to make a script to redirect us to TTY 1.

(I'm using GDM btw but i've seen this behaviour in sddm too...hope this helps)

**Comment by kRHYME7** [comment]
>So what we can do is to write a script that switches us to TTY 1 (login manager) in the process of switching we are never being actually logged out... and the fun fact is that if we have already logged in then the login manager will redirect us to the specific tty which is already logged in (instead of creating new). 


@Enigma1309  Hi have tried this on sddm and lightdm but both failed. If I go to the DM it will never respond because  the owner of it is the user of tty2. Now  I do not know how to install GDM as I might download all gnome dependencies. 

Now if we try to really implement this,  requirements might be the user should be using GDM, if GDM  also has a capability to to have multi user login sessions. Can you multi login using your DM ? 

I do not know what workarounds are available(Did not dig deeper about [multiseat](https://wiki.archlinux.org/title/xorg_multiseat) ).  I can only recommend is using the tty, polkit and chvt 


example If you want to login to  your tty3 run 
``` pkexec chvt 3 ```` the polkit will popup asking for sudo password then will change to tty3. Then Manually loging in to TTY 3 or using gettty service to autologin and autolauch user on tty3. which will make everything seemless .

the method above can be achieved by whatever DM is used.

**Comment by Enigma1309** [comment]
> Can you multi login using your DM ?

Yes it works without any problems, it even sends back to already logged in user if there is a session...

> Hi have tried this on sddm and lightdm but both failed

Okay i tried this on sddm and got the same result the SDDM seems to be freezing after loggin on..This may be the intended behaviour but that doesn't matter, the thing which matters is how to achieve this on other DM..

> I can only recommend is using the tty, polkit and chvt

Okay pkexec chvt 3 does change my tty but could we extend it like Automatically Launching the login manager To boot...This approach seems pretty good..

Also should we open a Feature Request in official Hyprland project so they might have a fix for it??

**Comment by kRHYME7** [comment]
Hello @Enigma1309   I added some useful features on the clipboard ``` Hyprdots show pastebin ```

It's not very conventional but should work as intended. I added Wipe and Delete options at the end of the list. 
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/db38e2a6-f038-4bca-ac4a-306e8c90aeeb)
Clicking wipe:
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/bce93e52-16bc-434a-89af-ef9827d40945)

Clicking Delete
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/bbb8b99d-ba96-4873-b5c9-cb531b1bb311)

By default ``` shift + return ``` lets you mark.
Here's the config I think you can change the keybind here : ~/.config/rofi/clipboard.rasi
Please refer to rofi manual on changing it.

**Comment by Enigma1309** [comment]
Thanks for the clipboard, i was active on issues and saw an issue by @exilonium on clipboard topic and i'm happily using it 🤗

as for user switching, i'm using gdm so what i do is manually switch to tty1 and choose whatever the user i want, it's easier  than implementing a script i guess 🙂

the only feature i miss is notification centre, hope there's a workaround..

**Comment by kRHYME7** [comment]
Tried making some script to make the tty change but ends up useless as I am used to changing via keyboard. 

About the notification-center, swaync is annoying as its syntax is not the same as Dunst so we can't keep both config. 

##### Already Tried a workaround: ( all notifications are on Dunst and only use the panel of swaync) 
+ If we switch to swaync, all pending notifications will be lost, when the notification already buffered to Dunst. (Done by killing Dunst process and use swaync daemon)

    The best solution I can think of is someone who can bang their head on writing eww or ags configuration to make a notification center (Or if you have time snatching someone's dots and maybe we can work on it 😄 ).

 While waiting for that hero, can you list down what specific features you would like to be included? 


BTW here's the progress, I stopped last month
I did not rice it im just experimenting if the workaround feels smooth
![image](https://github.com/prasanthrangan/hyprdots/assets/53417443/8280ab58-b325-47fc-a5ff-42829ce7513f)

**Comment by Enigma1309** [comment]
> Or if you have time snatching someone's dots and maybe we can work on it 😄 ).


i like [his](https://github.com/Aylur/dotfiles) dotfiles for notification centre, it's using ags

 > can you list down what specific features you would like to be included?

I am a simple guy i just want a basic place to access my old notification, a dnd mode, and media controls. 
switching to dnd should also buffer incoming notification.

I think we have to manually move away from dunst and choose notify-send command if we have to work with it..

**Comment by Enigma1309** [comment]
Hey !! any updates on this since the dependency of dunst was removed in the recent commit d64bed28a71955709029965f11238b0e6c63bcdd

**Comment by kRHYME7** [comment]
https://github.com/prasanthrangan/hyprdots/issues/1023 here let's just wait for now   I'm a li'l bit busy, so I can't contribute much.

**Comment by kRHYME7** [comment]
Try volume

**Comment by Enigma1309** [comment]
second notification was of  volume, i think you missed it probably

**Comment by kRHYME7** [comment]
oh my bad,  Let's just wait while tittu is cooking

**Comment by kRHYME7** [comment]
@Enigma1309  Can we close this for now in favor of https://github.com/prasanthrangan/hyprdots/issues/1023 (the one acknowledge by tittu) . Also, in case you got some other Feat Req, it would be fine if you'll push issues separately.

**Comment by Enigma1309** [comment]
yeah sure :)
